File: 0149-max-points-on-a-line/0149-max-points-on-a-line.py
class Solution:
    def findGCD(self, a, b):
        if b == 0:
            return a
        
        return self.findGCD(b, a%b)
    
    def maxPoints(self, points: List[List[int]]) -> int:
        
        N = len(points)
        
        if N == 1:
            return 1
        
        result = 0
        
        for i in range(N):
            
            freq = {}
            
            for j in range(N):
                
                if i == j:
                    continue
                
                dx = points[j][0] - points[i][0]
                dy = points[j][1] - points[i][1]
                
                gcd = self.findGCD(max(dx, dy), min(dx, dy))
                
                key = str(dx // gcd) + "_" + str(dy // gcd)
                
                freq[key] = freq.get(key, 0) + 1
            
            for key in freq.keys():
                result = max(result, freq[key]+1)
        
        return result
    

File: 0149-max-points-on-a-line/0149-max-points-on-a-line.java
class Solution {
    private int findGCD(int a, int b){
        
        if(b == 0) return a;
        
        return findGCD(b, a % b);
    }
    
    public int maxPoints(int[][] points) {
        
        int n = points.length;
        if(n == 1) return 1;
        
        int result = 0;
        
        for(int i=0; i<n; i++){
            
            Map<String, Integer> map = new HashMap<>();
            
            for(int j=0; j<n; j++){
                
                if(i == j) continue;
                
                int dx = points[j][0] - points[i][0];
                int dy = points[j][1] - points[i][1];
                
                int gcd = findGCD(Math.max(dx, dy), Math.min(dx, dy));
                
                String key = dx / gcd + "_" + dy / gcd;
                
                map.put(key, map.getOrDefault(key, 0) + 1);
            }
            
            for(String key : map.keySet()){
                result = Math.max(result, map.get(key) + 1);
            }
        }
        
        return result;
    }
}